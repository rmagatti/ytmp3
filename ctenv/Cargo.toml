[package]
name = "ctenv"
version = "0.1.0"
edition = "2021"

[lib]
proc-macro = true

[dependencies]
proc-macro2 = "1"
quote = "1"
syn = { version = "2", features = ["full"] }
dotenvy = "0.15"
proc-macro-error = "1.0.4"
[dev-dependencies]
trybuild = "1.0.105"
