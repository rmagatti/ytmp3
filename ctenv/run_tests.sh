#!/bin/bash

# Test runner script for ctenv

set -e

echo "=== Running ctenv Tests ==="

# Change to the ctenv directory
cd "$(dirname "$0")"

echo ""
echo "1. Running unit tests..."
cargo test --lib

echo ""
echo "2. Running integration tests..."
cargo test --test integration_tests

echo ""
echo "3. Running UI tests..."
cargo test --test ui_tests

echo ""
echo "4. Running dotenv integration tests..."
cargo test --test dotenv_integration

echo ""
echo "5. Running benchmark tests..."
cargo test --test benchmarks

echo ""
echo "6. Running examples..."
echo "6a. Comprehensive demo:"
cargo run --example comprehensive_demo

echo ""
echo "6b. Basic example from parent directory:"
(cd .. && RUST_LOG=info MY_CUSTOM_VAR=test_value cargo run --example ctenv_test)

echo ""
echo "7. Testing with different environment setups..."

# Test with specific environment variables set
echo "7a. Testing with compile-time environment variables:"
COMPILE_TIME_VAR="set_at_compile_time" cargo test test_compile_time_env_var

# Test with .env file
echo "7b. Testing with .env file:"
cp .env.test .env
cargo test test_runtime_fallback
rm -f .env

echo ""
echo "8. Building documentation..."
cargo doc --no-deps

echo ""
echo "=== All Tests Passed! ==="