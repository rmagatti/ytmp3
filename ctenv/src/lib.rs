use proc_macro::TokenStream;
use proc_macro_error::{proc_macro_error, emit_warning};
use quote::quote;
use syn::{parse_macro_input, LitStr};

/// A procedural macro that resolves environment variables at compile time when possible,
/// with runtime fallback.
///
/// Behavior:
/// 1. First tries to resolve from compile-time environment (e.g., `SUPABASE_URL=... cargo build`)
/// 2. Then tries to resolve from .env file at compile time
/// 3. If neither found, generates code that will call `std::env::var` at runtime
///
/// Usage: `ctenv!("ENVIRONMENT_VARIABLE_NAME")`
///
/// # Examples
///
/// ```rust,no_run
/// use ctenv::ctenv;
///
/// // This will be resolved at compile time if SUPABASE_URL is available,
/// // otherwise it will be resolved at runtime or panic if not set.
/// let supabase_url = ctenv!("SUPABASE_URL");
/// ```
#[proc_macro_error]
#[proc_macro]
pub fn ctenv(input: TokenStream) -> TokenStream {
    // Parse the key passed to the macro
    let var = parse_macro_input!(input as LitStr);
    let var_name = var.value();

    // 1. Real compile-time environment (cargo build SUPABASE_URL=...)
    if let Ok(val) = std::env::var(&var_name) {
        emit_warning!(
            var.span(),
            "found value for {} in environment at compile time",
            var_name
        );
        return quote!(#val.to_string()).into();
    }

    // 2. .env file present at compile time?
    if let Ok(iter) = dotenvy::from_path_iter(".env") {
        for item in iter.flatten() {
            if item.0 == var_name {
                emit_warning!(var.span(), "found value for {} in .env at compile time", var_name);
                let val = item.1;
                return quote!(#val.to_string()).into();
            }
        }
    }

    // 3. Fallback â€“ generate code that tries at run-time
    // (still panics if the var is missing then)
    emit_warning!(
        var.span(),
        "deferring resolution of {} to runtime",
        var_name
    );
    let tokens = quote! {
        ::std::env::var(#var_name)
             .expect(concat!("Environment variable ", #var_name, " is not set"))
    };
    tokens.into()
}
