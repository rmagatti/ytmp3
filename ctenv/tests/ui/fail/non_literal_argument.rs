// Test that macro fails with invalid argument type

use ctenv::ctenv;

fn main() {
    // Should fail - not a string literal
    let var_name = "SOME_VAR";
    let _var = ctenv!(var_name);
}