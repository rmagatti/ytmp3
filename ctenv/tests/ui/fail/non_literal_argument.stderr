error: expected string literal
 --> tests/ui/fail/non_literal_argument.rs:8:23
  |
8 |     let _var = ctenv!(var_name);
  |                       ^^^^^^^^
